cmd/download.go

```diff
@@ -176,6 +139,161 @@ func runDownload(cfg config.Config, flags *pflag.FlagSet, args []string) error {
+func newDownload(flags *pflag.FlagSet, usrCfg *viper.Viper) (*download, error) {
+	var err error
+	d := &download{}
+	d.uuid, err = flags.GetString("uuid")
+	if err != nil {
+		return nil, err
+	}
+	d.slug, err = flags.GetString("exercise")
+	if err != nil {
+		return nil, err
+	}
+	d.track, err = flags.GetString("track")
+	if err != nil {
+		return nil, err
+	}
+	d.team, err = flags.GetString("team")
+	if err != nil {
+		return nil, err
+	}
+
+	d.token = usrCfg.GetString("token")
+	d.apibaseurl = usrCfg.GetString("apibaseurl")
+	d.workspace = usrCfg.GetString("workspace")
+
+	if err = d.needsSlugXorUUID(); err != nil {
+		return nil, err
+	}
+	if err = d.needsUserConfigValues(); err != nil {
+		return nil, err
+	}
+	if err = d.needsSlugWhenGivenTrackOrTeam(); err != nil {
+		return nil, err
+	}
+
+	client, err := api.NewClient(d.token, d.apibaseurl)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := client.NewRequest("GET", d.url(), nil)
+	if err != nil {
+		return nil, err
+	}
+	d.buildPayloadQueryParams(req.URL)
+
+	res, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if err := json.NewDecoder(res.Body).Decode(&d.payload); err != nil {
+		return nil, fmt.Errorf("unable to parse API response - %s", err)
+	}
+
+	if res.StatusCode == http.StatusUnauthorized {
+		return nil, fmt.Errorf(
+			"unauthorized request. Please run the configure command. You can find your API token at %s/my/settings",
+			config.InferSiteURL(d.apibaseurl),
+		)
+	}
+	if res.StatusCode != http.StatusOK {
+		switch d.payload.Error.Type {
+		case "track_ambiguous":
+			return nil, fmt.Errorf(
+				"%s: %s",
+				d.payload.Error.Message,
+				strings.Join(d.payload.Error.PossibleTrackIDs, ", "),
+			)
+		default:
+			return nil, errors.New(d.payload.Error.Message)
+		}
+	}
+
+	return d, d.validate()
+}
+
+func (d download) url() string {
+	id := "latest"
+	if d.uuid != "" {
+		id = d.uuid
+	}
+	return fmt.Sprintf("%s/solutions/%s", d.apibaseurl, id)
+}
+
+// buildPayloadQueryParams adds optional query parameters to the URL.
+func (d download) buildPayloadQueryParams(url *netURL.URL) {
+	query := url.Query()
+	if d.slug != "" {
+		query.Add("exercise_id", d.slug)
+		if d.track != "" {
+			query.Add("track_id", d.track)
+		}
+		if d.team != "" {
+			query.Add("team_id", d.team)
+		}
+	}
+	url.RawQuery = query.Encode()
+}
+
+func (d download) validate() error {
+	if d.payload.Solution.ID == "" {
+		return errors.New("download missing ID")
+	}
+	if d.payload.Error.Message != "" {
+		return errors.New(d.payload.Error.Message)
+	}
+	return nil
+}
+
+// needsSlugXorUUID checks the presence of slug XOR uuid.
+func (d download) needsSlugXorUUID() error {
+	if d.slug != "" && d.uuid != "" || d.uuid == d.slug {
+		return errors.New("need an --exercise name or a solution --uuid")
+	}
+	return nil
+}
+
+// needsUserConfigValues checks the presence of required values from the user config.
+func (d download) needsUserConfigValues() error {
```

Comment:
In `cmd/cmd` we have `func validateUserConfig(cfg *viper.Viper)` but coupling to an outside validation seems bad so that's why it's duplicated here.